hypothesis-python/src/hypothesis/internal/conjecture/providers.py

```diff
@@ -0,0 +1,182 @@
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis/
+#
+# Copyright the Hypothesis Authors.
+# Individual contributors are listed in AUTHORS.rst and the git log.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+
+import math
+from typing import Optional
+
+from hypothesis.internal.compat import int_from_bytes
+from hypothesis.internal.conjecture.data import (
+    BYTE_MASKS,
+    COLLECTION_DEFAULT_MAX_SIZE,
+    ConjectureData,
+    PrimitiveProvider,
+    bits_to_bytes,
+)
+from hypothesis.internal.conjecture.floats import lex_to_float
+from hypothesis.internal.conjecture.utils import many
+from hypothesis.internal.floats import make_float_clamper
+from hypothesis.internal.intervalsets import IntervalSet
+
+
+class BytestringProvider(PrimitiveProvider):
+    lifetime = "test_case"
+
+    def __init__(
+        self, conjecturedata: Optional["ConjectureData"], /, *, bytestring: bytes
+    ):
+        super().__init__(conjecturedata)
+        self.bytestring = bytestring
+        self.index = 0
+        self.drawn = bytearray()
+
+    def _draw_bits(self, n):
+        if n == 0:  # pragma: no cover
+            return 0
+        n_bytes = bits_to_bytes(n)
+        if self.index + n_bytes > len(self.bytestring):
+            self._cd.mark_overrun()
+        buf = bytearray(self.bytestring[self.index : self.index + n_bytes])
+        self.index += n_bytes
+
+        buf[0] &= BYTE_MASKS[n % 8]
+        buf = bytes(buf)
+        self.drawn += buf
+        return int_from_bytes(buf)
+
+    def draw_boolean(
+        self,
+        p: float = 0.5,
+        *,
+        forced: Optional[bool] = None,
+        fake_forced: bool = False,
+    ) -> bool:
+        if forced is not None:
+            return forced
+
+        if p <= 0:
+            return False
+        if p >= 1:
+            return True
+
+        bits = math.ceil(-math.log2(min(p, 1 - p)))
+        # cap boolean draws at 8 bits so they are a constant size of one byte.
+        # If a probability requires more than 8 bits to precisely represent,
+        # the result will be slightly biased, but not badly.
+        bits = min(bits, 8)
+
+        size = 2**bits
+        falsey = math.floor(size * (1 - p))
```

Comment:
```suggestion
        falsey = max(1, math.floor(size * (1 - p)))
```
don't want to round low probabilities to zero!

Reply:
Also realized checking this that the bits calculation is very biased now that we don't recurse. I've replaced it with a constant 8 bits, which I believe is equivalent for p = {0.5, 0.75, ...} (only (top n) MSB matters <-> only `draw_bit(n)` matters) and is closer to the passed `p` otherwise.